src/mpid/ch4/netmod/ofi/ofi_send.h

```diff
@@ -345,14 +350,29 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_OFI_send(const void *buf, MPI_Aint count, MPI
         }
     } else {
         mpi_errno = MPIDI_OFI_send_normal(buf, count, datatype, cq_data, dst_rank, tag, comm,
-                                          context_offset, addr, request, dt_contig,
-                                          data_sz, dt_ptr, dt_true_lb, syncflag);
+                                          context_offset, addr, vni_src, vni_dst, request,
+                                          dt_contig, data_sz, dt_ptr, dt_true_lb, syncflag);
     }
 
     MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_OFI_SEND);
     return mpi_errno;
 }
 
+/* Common macro used by all MPIDI_NM_mpi_send routines to facilitate tuning */
+/* NOTE: concerned by the modulo? If we restrict num_vnis to power of 2,
+ * we may get away with bit mask */
+#define MPIDI_OFI_SEND_VNIS \
+    do { \
+        if (*request != NULL) { \
+            /* workq path */ \
+            vni_src = 0; \
+            vni_dst = 0; \
+        } else { \
+            vni_src = MPIDI_vci_get_src(comm, rank, tag) % MPIDI_OFI_global.num_vnis; \
+            vni_dst = MPIDI_vci_get_dst(comm, rank, tag) % MPIDI_OFI_global.num_vnis; \
+        } \
+    } while (0)
+
```

Comment:
Ugh... we spent a lot of time unrolling Charles's confusing no-argument macros. Please don't add more. Can this either be a `static inline` or at the very least, a macro with arguments to make it more clear what this is doing? When I was trying to read through the code, it really looked like `vni_src` was undefined.

Reply:
I understand. I thought I was defining and using this macro locally, within context it can be tolerated. Will adding `vni_src`, `vni_dst` to the macro argument make it better?

I haven't quite decided how optimized and how the workq code will finalize and having a macro avoids the potential typos and inconsistency which is often difficult to hunt down.

Reply:
Adding `vni_src` and `vni_dst` would clean things up enough for me.

Reply:
done.